### PR TITLE
[Spec] Apply custom logit processor in EAGLE-v2 verify path

### LINKED ISSUE #26330: Custom logit processor bypassed when speculative decoding is enabled
## Problem

When speculative decoding is enabled (e.g. `--speculative-algorithm NEXTN`), the custom logit processor registered via `--enable-custom-logit-processor` is not applied during token generation. The `thinking_budget` parameter in `custom_params` has no effect — the model generates reasoning tokens far beyond the configured budget.

Without speculative decoding, the same request with `thinking_budget: 1200` correctly limits reasoning to ~1200 tokens.

## Reproduction

**Server startup:**
```bash
sglang serve \
  --model-path /path/to/qwen3-fp8 \
  --trust-remote-code --quantization fp8 \
  --enable-custom-logit-processor \
  --reasoning-parser qwen3 \
  --speculative-algorithm NEXTN \
  --speculative-num-steps 5 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 6
```

**Request (curl, direct to SGLang):**
```bash
curl http://127.0.0.1:1234/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "qwen3",
    "messages": [{"role": "user", "content": "Prove that there are infinitely many primes using at least three different methods."}],
    "stream": true,
    "stream_options": {"include_usage": true},
    "custom_logit_processor": "{\"callable\": \"80049555000000000000008c2a73676c616e672e7372742e73616d706c696e672e637573746f6d5f6c6f6769745f70726f636573736f72948c225177656e33355468696e6b696e674275646765744c6f67697450726f636573736f729493942e\"}",
    "custom_params": {"thinking_budget": 1200}
  }'
```

The `custom_logit_processor` value is a hex-encoded pickle of:
```python
{"callable": "sglang.srt.sampling.custom_logit_processor.Qwen35ThinkingBudgetLogitProcessor"}
```

## Observed behavior

| Method | Speculative decoding | `reasoning_tokens` | `reasoning_chars` | Budget enforced? |
|--------|---------------------|-------------------|-------------------|------------------|
| curl (direct) | Disabled | ~1200 | ~3898 | Yes |
| curl (direct) | Enabled (NEXTN) | **~3848** | **~12352** | **No** |
| OpenAI SDK | Disabled | ~1200 | ~3898 | Yes |
| OpenAI SDK | Enabled (NEXTN) | **~2957-3112** | **~9281-11575** | **No** |

The HTTP request body was verified via client-side interception — `custom_logit_processor` and `custom_params` are present and correct in all cases. The server receives them but does not apply the logit processor when speculative decoding is active.

**Important:** This is not a token counting issue where draft tokens are counted but rejected. The actual `reasoning_chars` output (~9000-12000) far exceeds what 1200 tokens would produce (~3000-4000 chars), confirming that the logit processor is completely bypassed, not just miscounted.

## Expected behavior

The custom logit processor should be applied regardless of whether speculative decoding is enabled, enforcing the `thinking_budget` limit consistently.

## Root cause hypothesis

Speculative decoding generates draft tokens in batches (via the draft model) before verifying them with the target model. The custom logit processor is a per-token callback that modifies logits before sampling — it may be skipped during the draft phase, and the verification phase may not re-apply it, allowing budget-exceeding tokens through.

## Environment

- SGLang version: `0.5.13.dev20260525+ge86fdf3a3`
- Model: Qwen3 (FP8)
- GPU: RTX PRO 6000 Blackwell 96GB
- Python: 3.10

### PR DESCRIPTION
**Pull request by inin-zou**
## Motivation

When speculative decoding runs through the **v2 / SPEC_V2** verify path — which is the **default**, since `SGLANG_ENABLE_SPEC_V2` defaults to `True` — custom logit processors registered via `--enable-custom-logit-processor` are **silently bypassed**. For example, a `thinking_budget` limiter correctly caps reasoning tokens *without* speculative decoding, but has no effect *with* it.

Fixes #26330. (#21724 earlier reported the same root cause for the EAGLE-v2 path.)

## Root cause

`apply_custom_logit_processor` is applied in the v1 verify path (`eagle_info.py`), as well as in `dflash_info.py` and `ngram_info.py`. But `EagleVerifyInputV2Mixin.sample` (`eagle_info_v2.py`) applied penalties, grammar masks, and sampling **without ever calling it** — so the processor is dropped for every spec algorithm that routes through the v2 verify path.

## Modifications

- **`eagle_info_v2.py`**: apply the custom logit processor on the verify logits before penalties/sampling, mirroring the v1 path (`num_tokens_in_batch=self.draft_token_num`).
- **`test/registered/unit/spec/test_eagle_v2_custom_logit_processor.py`**: CPU regression test that drives `EagleVerifyInputV2Mixin.sample()` with a spy custom logit processor and asserts it is invoked on the verify logits.

## Test plan

The regression test passes **with** the fix and fails **without** it (verified on an RTX 4090):

```
WITH fix    (apply_custom_logit_processor present):  1 passed
WITHOUT fix (line reverted):                         1 failed
```

I also confirmed the call's batch-shape contract directly: `apply_custom_logit_processor` applied to a `(bs * draft_token_num, vocab)` tensor with `num_tokens_in_batch=draft_token_num` passes its internal shape assertion and modifies the expected rows.

I did not run a full server-mode end-to-end budget-enforcement test — happy to add one if preferred.

## Checklist

- [x] Format with pre-commit (isort, ruff, black, codespell all pass)
- [x] Unit test added (passes with the fix, fails without it)
- [ ] Accuracy / speed benchmarks — N/A (correctness fix; no kernel or model-forward change)

---
### CI States

Latest PR Test (Base): :x: [Run #26968569993](https://github.com/sgl-project/sglang/actions/runs/26968569993)
Latest PR Test (Extra): :x: [Run #26968569850](https://github.com/sgl-project/sglang/actions/runs/26968569850)

### CODE REVIEW

**Comment by inin-zou** [comment]
Closing in favor of #21799 (@jacobzhang22, opened 2026-04-01), which fixes the same root cause — `EagleVerifyInputV2Mixin.sample` not applying `apply_custom_logit_processor` in the SPEC_V2 verify path — and is further along (already reviewed and independently verified). I missed it because it references #21724 rather than #26330; apologies for the duplicate.

Worth noting for triage: #21799 should also resolve **#26330** (the NEXTN report), since NEXTN routes through the same v2 verify path (`SGLANG_ENABLE_SPEC_V2` defaults to True).
